contracts/Splitter.sol

```diff
@@ -19,8 +24,24 @@ contract Splitter {
   function split() external payable onlyAlice() {
     require(msg.value > 0, "Must split more than zero ether!");
     require(msg.value % 2 == 0, "The ether to be splitted must be even!");
+
     uint256 half = msg.value / 2;
-    bob.transfer(half);
-    carol.transfer(half);
+    bobBalance += half;
+    carolBalance += half;
+  }
+
+  function withdraw() external {
+    require(msg.sender == bob || msg.sender == carol, "Only Bob or Carol can withdraw!");
+    if (msg.sender == bob) {
+      require(bobBalance > 0, "Bob's balance must be grater than zero!");
+      uint256 tempBobBalance = bobBalance;
```

Comment:
This is the second time you `SLOAD` `bobBalance`. Can you save some gas?

Reply:
I want to set `bobBalance = 0;` before I `msg.sender.transfer(tempBobBalance);` because if the sender is a contract, it may cause a problem, do know how can I improve here, can you give me some hint? thanks

Reply:
Doing `=0` before the transfer is correct.
My point is that `require(bobBalance > 0` does an `SLOAD` and `tempBobBalance = bobBalance;` too. But it is the same value. So don't do 2 `SLOAD`s, do only one. I am sure you know what I mean.

Reply:
If I understand correctly when I read `bobBalance` from storage, I need to `SLOAD`. I can read once and check `require` later.

Reply:
Yes, read once, use many times.